---
title: "Sperm short RNA-Seq data analysis"
output:
  html_document:
    theme: cerulean
    df_print: paged
    toc: true
    toc_float: true
editor_options:
  chunk_output_type: console
---

# Library
```{r sperm-1, message=FALSE, warning=FALSE}
library(Biostrings)
library(plyr)
library(dplyr)
library(futile.logger)
library(BiocParallel)
library(data.table)
devtools::load_all("../../../../shortRNA")
devtools::document("../../../../shortRNA")
```

# `fastq` files
```{r sperm-2 }
counts_file <- "fastq/counts.rds"

if (!file.exists(counts_file)) {
  files <- list.files(path = "./data", pattern = "gz", full.names = T)
  names(files) <- gsub(pattern = ".*/|_R1.*", replacement = "", x = files)

  m <- fastq2SeqCountMatrix(files = files)
  saveRDS(object = m, file = counts_file)
} else {
  m <- readRDS(counts_file)
}

fasta_file <- "fastq/unique.fasta"

if (!file.exists(fasta_file)) {
  fa <- DNAStringSet(row.names(m))
  # names(fa) <- paste0("S", 1:length(fa))
  names(fa) <- fa
  writeXStringSet(fa, fasta_file)
} else {
  fa <- readAAStringSet(fasta_file)
}
```

# Annotations
```{r sperm-3 }
anno_file <- "../genome/features.rds"
if (!file.exists(anno_file)) {
  db_mmu <- getDB()

  mm10_annoprep <- prepareAnnotation(
    ensdb = db_mmu$ensdb,
    genome = "/mnt/IM/reference/genome/gencode/fasta/GRCm38.p5.genome.fa",
    output_dir = "../genome",
    extra.gr = list(piRNA = db_mmu$piRNA_GR, miRNA = db_mmu$miRNA_GR),
    extra.seqs = list(rRNA = db_mmu$rRNA_fa, tRNA = db_mmu$tRNA_fa),
    resolveSplicing = NULL,
    rules = defaultAssignRules(),
    tRNAEnsembleRemove = TRUE,
    clusterMiRNA = TRUE
  )

  saveRDS(db_mmu, "./annotation/db.rds")
} else {
  a <- readRDS(anno_file)
  db <- readRDS("./annotation/db.rds")
}
```


# Align data
```{r sperm-5 }
align_file <- "align/unique.bam"
if (!file.exists(align_file)) {
  alignShortRNA(
    fastq = "fastq/unique.fasta",
    index = "../../../../shortRNA/genome/customGenome",
    outDir = "align", GTF = exonsBy(db$ensdb),
    GTF.featureType = "exon", GTF.attrType = "gene_id"
  )
} else {
  cat("Alignment file exists!\n")
}
```

```
        ==========     _____ _    _ ____  _____  ______          _____  
        =====         / ____| |  | |  _ \|  __ \|  ____|   /\   |  __ \ 
          =====      | (___ | |  | | |_) | |__) | |__     /  \  | |  | |
            ====      \___ \| |  | |  _ <|  _  /|  __|   / /\ \ | |  | |
              ====    ____) | |__| | |_) | | \ \| |____ / ____ \| |__| |
        ==========   |_____/ \____/|____/|_|  \_\______/_/    \_\_____/
       Rsubread 2.4.3

//================================= setting ==================================\\
||                                                                            ||
|| Function      : Read alignment (RNA-Seq)                                   ||
|| Input file    : unique.fasta                                               ||
|| Output file   : unique.bam (BAM), Sorted                                   ||
|| Index name    : customGenome                                               ||
||                                                                            ||
||                    ------------------------------------                    ||
||                                                                            ||
||                               Threads : 16                                 ||
||                          Phred offset : 33                                 ||
||                             Min votes : 3 / 10                             ||
||                        Max mismatches : 3                                  ||
||                      Max indel length : 5                                  ||
||            Report multi-mapping reads : yes                                ||
|| Max alignments per multi-mapping read : 100                                ||
||                                                                            ||
\\============================================================================//

//================ Running (01-Jul-2021 13:16:49, pid=19931) =================\\
||                                                                            ||
|| Check the input reads.                                                     ||
|| The input file contains base space reads.                                  ||
|| Initialise the memory objects.                                             ||
|| Estimate the mean read length.                                             ||
|| Create the output BAM file.                                                ||
|| Check the index.                                                           ||
|| Init the voting space.                                                     ||
|| Global environment is initialised.                                         ||
|| Load the 1-th index block...                                               ||
|| The index block has been loaded.                                           ||
|| Start read mapping in chunk.                                               ||
||    3% completed, 0,7 mins elapsed, rate=109,4k reads per second            ||
||   10% completed, 0,7 mins elapsed, rate=124,9k reads per second            ||
||   16% completed, 0,7 mins elapsed, rate=130,0k reads per second            ||
||   23% completed, 0,7 mins elapsed, rate=131,9k reads per second            ||
||   29% completed, 0,7 mins elapsed, rate=134,1k reads per second            ||
||   36% completed, 0,8 mins elapsed, rate=135,4k reads per second            ||
||   43% completed, 0,8 mins elapsed, rate=136,0k reads per second            ||
||   50% completed, 0,8 mins elapsed, rate=135,9k reads per second            ||
||   56% completed, 0,8 mins elapsed, rate=136,0k reads per second            ||
||   63% completed, 0,9 mins elapsed, rate=136,0k reads per second            ||
||   70% completed, 0,9 mins elapsed, rate=35,3k reads per second             ||
||   73% completed, 0,9 mins elapsed, rate=36,2k reads per second             ||
||   77% completed, 0,9 mins elapsed, rate=37,4k reads per second             ||
||   81% completed, 0,9 mins elapsed, rate=38,4k reads per second             ||
||   84% completed, 1,0 mins elapsed, rate=39,3k reads per second             ||
||   88% completed, 1,0 mins elapsed, rate=40,2k reads per second             ||
||   91% completed, 1,0 mins elapsed, rate=41,0k reads per second             ||
||   93% completed, 1,0 mins elapsed, rate=41,6k reads per second             ||
||   95% completed, 1,0 mins elapsed, rate=42,0k reads per second             ||
||   99% completed, 1,0 mins elapsed, rate=42,8k reads per second             ||
||                                                                            ||
||                           Completed successfully.                          ||
||                                                                            ||
\\====================================    ====================================//

//================================   Summary =================================\\
||                                                                            ||
||                 Total reads : 2.689.581                                    ||
||                      Mapped : 727.577 (27,1%)                              ||
||             Uniquely mapped : 454.332                                      ||
||               Multi-mapping : 273.245                                      ||
||                                                                            ||
||                    Unmapped : 1.962.004                                    ||
||                                                                            ||
||                      Indels : 1.211                                        ||
||                                                                            ||
||                Running time : 1,2 minutes                                  ||
||                                                                            ||
\\============================================================================//

                      unique.bam
Total_reads              2689581
Mapped_reads              727577
Uniquely_mapped_reads     454332
Multi_mapping_reads       273245
Unmapped_reads           1962004
Indels                      1211
```


# get the overlaps between alignments and annotation

```{r sperm-6 }
overlap_file <- "align/overlapBAM.rds"
if (!file.exists(overlap_file)) {
  o <- overlapWithTx2(
    bamFile = align_file, annotation = a,
    ignoreStrand = TRUE, nbthreads = 16
  )
  saveRDS(object = o, file = overlap_file)
} else {
  o <- readRDS(overlap_file)
}
```

```
1154610 alignments loaded, searching for overlaps...
Found 790751 overlaps.
Calculating positions relative to transcripts...
Aggregating...
Warning messages:
1: In .Seqinfo.mergexy(x, y) :
  Each of the 2 combined objects has sequence levels not in the other:
  - in 'x': GL456213.1, GL456359.1, GL456360.1, GL456366.1, GL456367.1, GL456368.1, GL456370.1, GL456378.1, GL456379.1, GL456382.1, GL456383.1, GL456387.1, GL456389.1, GL456390.1, GL456392.1, GL456393.1, GL456394.1, GL456396.1, JH584300.1, JH584301.1, JH584302.1, pseudoChr_SSU-AK173067(possible, pseudoChr_SSU-JC106267(possible, pseudoChr_SSU-CAAA01088223(possible, pseudoChr_SSU-CAAA01160619(possible, pseudoChr_LSU-GU372691(28S
  - in 'y': CHR_CAST_EI_MMCHR11_CTG4, CHR_CAST_EI_MMCHR11_CTG5, CHR_MG104_PATCH, CHR_MG117_PATCH, CHR_MG132_PATCH, CHR_MG153_PATCH, CHR_MG171_PATCH, CHR_MG184_PATCH, CHR_MG190_MG3751_PATCH, CHR_MG191_PATCH, CHR_MG209_PATCH, CHR_MG3172_PATCH, CHR_MG3231_PATCH, CHR_MG3251_PATCH, CHR_MG3490_PATCH, CHR_MG3496_PATCH, CHR_MG3530_PATCH, CHR_MG3561_PATCH, CHR_MG3562_PATCH, CHR_MG3609_PATCH, CHR_MG3618_PATCH, CHR_MG3627_PATCH, CHR_MG3648_PATCH, CHR_MG3656_PATCH, CHR_MG3683_PATCH, CHR_MG3686_PATCH, CHR_MG [... truncated]
2: In .Seqinfo.mergexy(x, y) :
  Each of the 2 combined objects has sequence levels not in the other:
  - in 'x': GL456213.1, GL456359.1, GL456360.1, GL456366.1, GL456367.1, GL456368.1, GL456370.1, GL456378.1, GL456379.1, GL456382.1, GL456383.1, GL456387.1, GL456389.1, GL456390.1, GL456392.1, GL456393.1, GL456394.1, GL456396.1, JH584300.1, JH584301.1, JH584302.1, pseudoChr_SSU-AK173067(possible, pseudoChr_SSU-JC106267(possible, pseudoChr_SSU-CAAA01088223(possible, pseudoChr_SSU-CAAA01160619(possible, pseudoChr_LSU-GU372691(28S
  - in 'y': CHR_CAST_EI_MMCHR11_CTG4, CHR_CAST_EI_MMCHR11_CTG5, CHR_MG104_PATCH, CHR_MG117_PATCH, CHR_MG132_PATCH, CHR_MG153_PATCH, CHR_MG171_PATCH, CHR_MG184_PATCH, CHR_MG190_MG3751_PATCH, CHR_MG191_PATCH, CHR_MG209_PATCH, CHR_MG3172_PATCH, CHR_MG3231_PATCH, CHR_MG3251_PATCH, CHR_MG3490_PATCH, CHR_MG3496_PATCH, CHR_MG3530_PATCH, CHR_MG3561_PATCH, CHR_MG3562_PATCH, CHR_MG3609_PATCH, CHR_MG3618_PATCH, CHR_MG3627_PATCH, CHR_MG3648_PATCH, CHR_MG3656_PATCH, CHR_MG3683_PATCH, CHR_MG3686_PATCH, CHR_MG [... truncated]
3: In .Seqinfo.mergexy(x, y) :
  Each of the 2 combined objects has sequence levels not in the other:
  - in 'x': GL456213.1, GL456359.1, GL456360.1, GL456366.1, GL456367.1, GL456368.1, GL456370.1, GL456378.1, GL456379.1, GL456382.1, GL456383.1, GL456387.1, GL456389.1, GL456390.1, GL456392.1, GL456393.1, GL456394.1, GL456396.1, JH584300.1, JH584301.1, JH584302.1, pseudoChr_SSU-AK173067(possible, pseudoChr_SSU-JC106267(possible, pseudoChr_SSU-CAAA01088223(possible, pseudoChr_SSU-CAAA01160619(possible, pseudoChr_LSU-GU372691(28S
  - in 'y': CHR_CAST_EI_MMCHR11_CTG4, CHR_CAST_EI_MMCHR11_CTG5, CHR_MG104_PATCH, CHR_MG117_PATCH, CHR_MG132_PATCH, CHR_MG153_PATCH, CHR_MG171_PATCH, CHR_MG184_PATCH, CHR_MG190_MG3751_PATCH, CHR_MG191_PATCH, CHR_MG209_PATCH, CHR_MG3172_PATCH, CHR_MG3231_PATCH, CHR_MG3251_PATCH, CHR_MG3490_PATCH, CHR_MG3496_PATCH, CHR_MG3530_PATCH, CHR_MG3561_PATCH, CHR_MG3562_PATCH, CHR_MG3609_PATCH, CHR_MG3618_PATCH, CHR_MG3627_PATCH, CHR_MG3648_PATCH, CHR_MG3656_PATCH, CHR_MG3683_PATCH, CHR_MG3686_PATCH, CHR_MG [... truncated]
```


# We assign the reads
```{r sperm-9 }
assigned_file <- "align/assignedReads.rds"

if (!file.exists(assigned_file)) {
  ar <- assignReads(sources = o, rules = defaultAssignRules())
  saveRDS(object = ar, file = assigned_file)
} else {
  ar <- readRDS(assigned_file)
}
```


# Annotatins factorList
```{r}
fl_file <- "annotation/annoFL.rds"

if (!file.exists(assigned_file)) {
  fl <- featuresAnnoToFL(a)
  names(fl) <- lapply(fl, function(x) as.character(x[length(x)]))
  saveRDS(object = fl, file = fl_file)
} else {
  fl <- readRDS(fl_file)
}
```

# Tree construction
```{r}
tree_file <- "annotation/tree.rds"
if (!file.exists(tree_file)) {
  mappedFeaturesDF <- ar
  mappedFeaturesDF$seq <- rownames(mappedFeaturesDF)
  ar_tree <- addReadsToTree(
    fL = fl,
    mappedFeaturesDF = mappedFeaturesDF,
    unassigned = FALSE,
    extraTreeBranch = NULL
  )
  saveRDS(object = ar_tree, file = tree_file)
} else {
  ar_tree <- readRDS(tree_file)
}

# write.tree(ar_tree, file = "annotation/tree.tree", append = FALSE, digits = 10, tree.names = FALSE)
```


# TSE construction
```{r}
library(TreeSummarizedExperiment)

rt <- ar_tree
as <- list(counts = m)

tse <- TreeSummarizedExperiment(
  assays = as,
  rowTree = rt,
  metadata = list(
    assignedReads = ar,
    counts = m,
    notAligned = getReadsFromBam(
      bam = align_file
    )
  )
)

saveRDS(object = tse, file = "tse.rds", compress = T)
```


# Playing with tree
```{r}
library(ape)
library(treeio)
library(ggplot2)
library(ggtree)
library(data.tree)

# dt <- data.tree::as.Node(ar_tree)

art_tibble <- as_tibble(ar_tree)

art_treeData <- as.treedata(ar_tree)

# art_subset <- tree_subset(art_treeData, "tRNA")

# art_subset1 <- tree_subset(ar_tree, "let7")

summary(ar_tree)

library(castor)
tree <- get_subtree_at_node(ar_tree, "tRNA-Arg-ACG-1")$subtree

p <- ggtree(tree, color = "firebrick", size = 0.5, branch.length = "none") +
  geom_nodelab(size = 5, color = "purple", angle = 90, vjust = -0.5) +
  geom_tiplab(size = 5, color = "black", as_ylab = TRUE)

viewClade(p, MRCA(p, "tRNA-Arg-ACG-1-2"))

gheatmap(p, m)
```
